required/tools/l3sys-query.dtx

```diff
@@ -0,0 +1,459 @@
+% information is available in modern \TeX{} systems using primitives, and partly
+% as the aim of \texttt{l3sys-query} is to provide information where there are
+% defined use cases. Requests for additional data interfaces are welcome.
+%
+% This `package provides a document level \textit{Key-Value} interface
+% to The \texttt{l3sys-query} functionality, although as a convenience
+% it also summarizes the documentation for the L3 programming
+% interface layer (prvided by the \pkg{l3sys} module of
+% \pkg{l3kernel}, and of the command line Lua script
+% \texttt{l3sys-query} which is distributed separately and provides
+% the underlying functionality.
+%
+% 
+% 
+% \section{The document level key-value  interface\label{sec:kv}}
+%
+% The package just provides three commands.
+%
+% \begin{function}
+%   {\QueryWorkingDirectory}
+%   \begin{syntax}
+%     \cs{QueryWorkingDirectory} \Arg{result cmd}
+%   \end{syntax}
+%   Returns the current working directory of the \TeX\ system in the
+%   supplied command \meta{result cmd}.
+% \end{function}
+% 
+% \begin{function}
+%   {\QueryFiles, \QueryFilesTF}
+%   \begin{syntax}
+%     \cs{QueryFiles} \oarg{options} \Arg{spec} \Arg{function}
+%     \cs{QueryFilesTF} \oarg{options} \Arg{spec} \Arg{function} \Arg{pre code} \Arg{empty list code}
+%   \end{syntax}
+%   Returns the current working directory of the \TeX\ system in the
+%   supplied command \meta{result cmd}.
+% \end{function}
+%
+% \section{The L3 programming layer interface\label{sec:expl3}}
+% 
+% \begin{function}
+%   {\sys_get_query:nN, \sys_get_query:nnN, \sys_get_query:nnnN}
+%   \begin{syntax}
+%     \cs{sys_get_query:nN} \Arg{cmd} \Arg{tl var}
+%     \cs{sys_get_query:nnN} \Arg{cmd} \Arg{spec} \Arg{tl var}
+%     \cs{sys_get_query:nnnN} \Arg{cmd} \Arg{options} \Arg{spec} \Arg{tl var}
+%   \end{syntax}
+%   Sets the \meta{tl var} to the information returned by the
+%   \texttt{l3sys-query} \meta{cmd}, potentially supplying the \meta{options}
+%   and \meta{spec} to the query call. The valid \meta{cmd} names are at present
+%   \begin{itemize}
+%     \item \texttt{pwd} Returns the present working directory
+%     \item \texttt{ls} Returns a directory listing, using the \meta{spec} to
+%       select files and applying the \meta{options} if given
+%   \end{itemize}
+%   The \meta{spec} should be a file glob and will automatically be passed to
+%   the script without shell expansion.
+% \end{function}
+% 
+% \begin{function}
+%   {\sys_split_query:nN, \sys_split_query:nnN, \sys_split_query:nnnN}
+%   \begin{syntax}
+%     \cs{sys_split_query:nN} \Arg{cmd} \Arg{seq}
+%     \cs{sys_split_query:nnN} \Arg{cmd} \Arg{spec} \Arg{seq}
+%     \cs{sys_split_query:nnnN} \Arg{cmd} \Arg{options} \Arg{spec} \Arg{seq}
+%   \end{syntax}
+%   Works as described for \cs{sys_split_query:nnnN}, but sets the \meta{seq}
+%   to contain one entry for each line returned by \texttt{l3sys-query}.
```

Comment:
that sounds like a recursive definition or am I missing something?  \sys_split_query:nnnN works like \sys_split_query:nnnN but ...

Reply:
yes (same text in the l3sys code doc) should refer to `\sys_get_query:nnnN` that returns a tl, this is same but returns a seq, will fix